include/RAJA/exec-cuda/forallN_cuda.hxx

```diff
@@ -139,31 +139,34 @@ struct CudaIndexPair : public POL {
   typedef IDX INDEX;
 };
 
+/** Provides a range from 0 to N_iter - 1
+ * 
+ */
 template <typename VIEWDIM, int threads_per_block>
 struct CudaThreadBlock {
-  int begin;
-  int end;
+  int distance;
 
   VIEWDIM view;
 
-  CudaThreadBlock(CudaDim &dims, RangeSegment const &is)
-      : begin(is.getBegin()), end(is.getEnd())
+  template<typename Iterable>
+  CudaThreadBlock(CudaDim &dims, Iterable const &i)
+      : distance(std::distance(std::begin(i), std::end(i)))
   {
     setDims(dims);
   }
 
   __device__ inline int operator()(void)
   {
-    int idx = begin + view(blockIdx) * threads_per_block + view(threadIdx);
-    if (idx >= end) {
+    int idx = 0 + view(blockIdx) * threads_per_block + view(threadIdx);
```

Comment:
It's been turned a bit inside-out.  Instead of storing begin and end, which may be meaningless, it stores the number of iterator results between begin and end and gives each thread one of those.  Those are then passed to `iterator::operator[]` by the new CudaIterableWrapper class, which should return whatever iterator value is at that offset from begin.